Output build dependencies of a specific image repository

kargakis opened the pull request:
Supported output formats: json, dot, and ast.
```
$ openshift ex build-chain ruby-20-centos7
{
	"fullname": "default/ruby-20-centos7",
	"tags": [
		""
	],
	"edges": [
		{
			"fullname": "default/ruby-sample-build",
			"to": "default/origin-ruby-sample"
		}
	],
	"children": [
		{
			"fullname": "default/origin-ruby-sample",
			"tags": [
				""
			]
		}
	]
}

$ openshift ex build-chain ruby-20-centos7 -o dot
digraph ruby_20_centos7 {
	ruby_20_centos7->origin_ruby_sample[ label=<ruby_sample_build<BR /><FONT POINT-SIZE="10">default</FONT>> ];
	origin_ruby_sample [ comment="", label=<origin_ruby_sample<BR /><FONT POINT-SIZE="10">default</FONT>> ];
	ruby_20_centos7 [ comment="", label=<ruby_20_centos7<BR /><FONT POINT-SIZE="10">default</FONT>> ];

}

$ openshift ex build-chain ruby-20-centos7 -o ast
ruby-20-centos7(origin-ruby-sample)
```

DOT output visualized
http://i.imgur.com/LVse6hX.png

smarterclayton commented:
We should have three formats:

internally, in memory, a structure that represents the data
externally, a versioned JSON format that can be dumped
externally, a graphviz transformation of the in memory format.

the first two should be very closely coupled - they don't have to be formal API objects, but it would be better to structure them that way so that in the future we can version them.  I would think of the graphviz as a special "-o" argument that visualizes them (-o viz)

----- Original Message -----
> cc: @csrwng @smarterclayton @bparees
> 
> Opening this issue to sync with
> [this](https://trello.com/c/52pEs31Q/510-add-command-line-utility-to-output-build-dependencies-of-a-specific-image-repository)
> Trello card.
> 
> @smarterclayton I was thinking of going straight to the dot stuff. There is a
> library that can make things very easy:
> https://godoc.org/code.google.com/p/gographviz
> What do you think?
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/issues/1229

kargakis commented:
@bparees @csrwng ptal

I have created the basic structure of the code (data structures, flow, etc). I haven't thoroughly tested it though except starting two builds from the same configuration file just to see if it can find the pushed images (and it does).

```
[vagrant@openshiftdev sample-app]$ openshift cli get images
NAME                DOCKER REF
[vagrant@openshiftdev sample-app]$ openshift cli get buildConfigs
NAME                TYPE                SOURCE
ruby-sample-build   STI                 git://github.com/openshift/ruby-hello-world.git
[vagrant@openshiftdev sample-app]$ openshift cli start-build ruby-sample-build
ruby-sample-build-422f94ba-c407-11e4-9426-080027c5bfa9
[vagrant@openshiftdev sample-app]$ openshift cli start-build ruby-sample-build
ruby-sample-build-5a35194d-c408-11e4-9426-080027c5bfa9
[vagrant@openshiftdev sample-app]$ openshift cli get images
NAME                                                               DOCKER REF
26393db67314c93f2a42ef2f2cc6266b872a55443fc57f6a126d48d31a6b74b3   172.30.17.83:5000/default/origin-ruby-sample:26393db67314c93f2a42ef2f2cc6266b872a55443fc57f6a126d48d31a6b74b3
ff5ea1a3e3c7dd7ea8869cdee53be964a2460357b313bc724aecf3aa6440e432   172.30.17.83:5000/default/origin-ruby-sample:ff5ea1a3e3c7dd7ea8869cdee53be964a2460357b313bc724aecf3aa6440e432
[vagrant@openshiftdev sample-app]$ openshift ex build-chain -i origin-ruby-sample
Image repo origin-ruby-sample has 2 image(s)
```

For testing all of the code, I have to create multiple Dockerfiles that depend on each other, right? :soundsboring: :)

bparees commented:
@kargakis Well this is an interesting question...   do we want to illustrate
1) any build that would be triggered by an upstream image change, even if that build doesn't explicitly consume the upstream image
or
2) only the cases where the build actually consumes the upstream image?

I think the (1) is sufficient, and given that, you don't need do anything w/ actual dockerfiles..you can just create a set of BuildConfigs that have some ImageChangeTrigger relationship.  It doesn't really matter if the actual BuildConfig consumes the new/changed image or not, just that the upstream image changing causes that BuildConfig to run (which in turn produces some new image that may or may not actually be based on the upstream image)

kargakis commented:
> 1) any build that would be triggered by an upstream image change, even if that build doesn't explicitly consume the upstream image

I don't understand this. How can a build be triggered by an upstream image if that image is not a part of the build chain (ie. one of the images that can be found in FROM up in the tree) ?

bparees commented:
@kargakis nothing prevents me from setting up an ImageChangeTrigger on a build in which the ICT watches image "foo" for changes but the BuildConfig itself does not consume "foo" as part of the build process.

bparees commented:
@kargakis similarly you could setup a Github webhook that triggered your build on pushes to repo Foo even though your build was building repo Bar

kargakis commented:
Ok, so to make this clear I am going to check for triggers in configuration files **as well as** images that explicitly depend on other images (which I already do), right?

bparees commented:
@kargakis no, just triggers.  @smarterclayton @csrwng please confirm.

csrwng commented:
@bparees @kargakis - that's my understanding. Only triggers.

kargakis commented:
Until deciding how are we going to look for dependencies (see Jay's comment on Trello), I've started writting tests for parts of the code that are independent [of the way we inspect for dependencies] . Most important being the building of the tree structure (`buildTree`).

kargakis commented:
When running `go test` in the buildchain directory I get the following error message:

```
$ cd pkg/cmd/experimental/buildchain/
$ go test
# github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation
../../../../../../GoogleCloudPlatform/kubernetes/pkg/api/validation/schema.go:129: invalid operation: fieldDetails.DataTypeFields.Items[0] (type *swagger.Item does not support indexing)
FAIL	github.com/kargakis/origin/pkg/cmd/experimental/buildchain [build failed]
```

Is anybody familair with it? Tests should pass. Here is a slightly edited version of the algorithm (and tests) using strings instead of api.Images: http://play.golang.org/p/qEo_nGt5Ko

smarterclayton commented:
I cleared this up with jay - the approach Ben relayed is the one we are doing for now.  


> On Mar 9, 2015, at 8:12 AM, Michail Kargakis <notifications@github.com> wrote:
> 
> Until deciding how are we going to look for dependencies (see Jay's comment on Trello), I've started writting tests for parts of the code that are independent [of the way we inspect for dependencies] . Most important being the building of the tree structure (buildTree).
> 
> —
> Reply to this email directly or view it on GitHub.
>

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,164 @@
+
+// isValid checks if the given image repository is valid
+func isValid(name string, c *client.Client, namespace string) bool {
+	// TODO: Validate image repository via buildConfigs?
```
Since we probably haven't run any buildConfigs, these image repositories do not exist, so anything that goes here will be baptized as invalid. So I think we should check for image repositories in buildConfigs right? 

(cc: @bparees @csrwng )

csrwng replied:
The image repositories will exist. Only the images/tags won't.

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,164 @@
+	for _, child := range root.children {
+		// Search through tags
+		for _, tag := range child.tags {
+			// TODO: These images probably do not exist since we haven't run any buildConfigs
```
The same as above with the repos, these images do not exist.

(cc: @bparees @csrwng )

csrwng replied:
Correct... which is why I think we should only be looking at the repo relationships (to)

smarterclayton replied:
Right- we should output the other value, and maybe in a later pull try to infer overlap / reductions


> On Mar 10, 2015, at 11:51 AM, Cesar Wong <notifications@github.com> wrote:
> 
> In pkg/cmd/experimental/buildchain/buildchain.go:
> 
> > +
> > +// repoToImage accepts an image repository tree and returns its equivalent
> > +// image tree. The root node keeps on representing the image repository
> > +// specified with the --imagerepo (-i) flag though
> > +func repoToImage(c *client.Client, namespace string, nodeName string, root *node) *node {
> > +	if root == nil {
> > +		return nil
> > +	}
> > +	nd := &node{
> > +		name: nodeName,
> > +	}
> > +
> > +	for _, child := range root.children {
> > +		// Search through tags
> > +		for _, tag := range child.tags {
> > +			// TODO: These images probably do not exist since we haven't run any buildConfigs
> Correct... which is why I think we should only be looking at the repo relationships (to)
> 
> —
> Reply to this email directly or view it on GitHub.
>

kargakis replied:
Cool, just image repository trees it is then.

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,164 @@
+		for _, trigger := range cfg.Triggers {
+			if from == trigger.ImageChange.From.Name {
+				// Search recursively all children for their dependency trees
+				// TODO: Check for both outputs ? (To & Tag or DockerImageReference)
```
(cc: @bparees @csrwng )

csrwng replied:
I would think we're only interested in output.to, @bparees ?

smarterclayton replied:
We're interested in both, but they mean different things.

----- Original Message -----
> > +
> > +// findRepoDeps accepts an image repository name and a list of build
> > +// configurations and returns a dependency tree of the specified
> > +// image repository along with its tags
> > +func findRepoDeps(from string, buildConfigList []buildapi.BuildConfig)
> > *node {
> > +	root := &node{
> > +		name: from,
> > +	}
> > +
> > +	// Search all build configurations in order to find the image
> > +	// repositories depending on the specified image repository
> > +	for _, cfg := range buildConfigList {
> > +		for _, trigger := range cfg.Triggers {
> > +			if from == trigger.ImageChange.From.Name {
> > +				// Search recursively all children for their dependency trees
> > +				// TODO: Check for both outputs ? (To & Tag or DockerImageReference)
> 
> I would think we're only interested in output.to, @bparees ?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26134297

kargakis commented:
Are we addressing circular dependencies in buildConfigs atm?

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,162 @@
+		Run: func(cmd *cobra.Command, args []string) {
+			imageRepoName := cmdutil.GetFlagString(cmd, "imagerepo")
+			if imageRepoName == "" {
+				usageError(cmd, "Must pass an image repository name with the --imagerepo (-i) flag")
```
Shouldn't you just take the first argument as the image repo name? (line 45 implies that you do)

smarterclayton replied:
There should be a mode for this command that is --all that iterates over all image repositories in the system and creates a graph with multiple roots.

----- Original Message -----
> > +Output build dependencies of a specific image repository.
> > +
> > +Examples:
> > +
> > +    $ openshift ex build-chain [image-repository]
> > +`
> > +
> > +func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string)
> > *cobra.Command {
> > +	cmd := &cobra.Command{
> > +		Use:   fmt.Sprintf("%s [image-repository]", name),
> > +		Short: "Output build dependencies of a specific image repository",
> > +		Long:  longDescription,
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			imageRepoName := cmdutil.GetFlagString(cmd, "imagerepo")
> > +			if imageRepoName == "" {
> > +				usageError(cmd, "Must pass an image repository name with the
> > --imagerepo (-i) flag")
> 
> Shouldn't you just take the first argument as the image repo name? (line 45
> implies that you do)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26130209

kargakis replied:
Good catch, I'll just use the first argument.

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// isValid checks if the given image repository is valid
+func isValid(name string, c *client.Client, namespace string) bool {
```
Just a matter of style... I think name and namespace should likely be next to each other.

bparees commented:
@kargakis no, we are not.

kargakis commented:
> @kargakis no, we are not.

@bparees fyi I just created a Trello card for this

kargakis commented:
JSON is in, next thing tbd DOT

kargakis commented:
@bparees @csrwng @smarterclayton ptal

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,210 @@
+					_, _, name, tag, err = imageapi.SplitDockerPullSpec(cfg.Parameters.Output.DockerImageReference)
+					checkErr(err)
+				} else {
+					name = cfg.Parameters.Output.To.Name
```
I believe it's valid for Output.To to specify a namespace (Output.To.Namespace) so this should take that into account.  (if not specified it defaults to this namespace)

kargakis replied:
@bparees the same goes for the [DockerImageReference](https://github.com/openshift/origin/blob/5ba08eb6f4ec7a18397b404aa821f84073d4d493/pkg/image/api/helper.go#L17), right?

bparees replied:
no, DockerImageReference is a docker pull spec for a registry, it has no association to an openshift namespace.

bparees commented:
I think we need to specify a tag when running this, because a particular ICT isn't triggered just by an ImageRepo changing, the specific Tag that the ICT watches needs to change.

kargakis commented:
> I think we need to specify a tag when running this, because a particular ICT isn't triggered just by an ImageRepo changing, the specific Tag that the ICT watches needs to change.

The idea is that the user specifies an image repository and gets back a dependency tree. Now, the tags you are talking about can be found in the tree, ie. each node has alongside its name and namespace, the tags by which the node depends on another node.

kargakis commented:
The problem that rises is that if one repository A depends on more than one repos, then currently we don't correlate repo's A tags with the rest repos.

bparees commented:
if my build is triggered by an update to imageFoo:v2 and i'm about to change imageFoo:v3, I don't want to query "show me everything that's related to ImageFoo", I want to query "show me everything related to ImageFoo:v3"

bparees commented:
Everything downstream of ImageFoo:V3 really.

kargakis commented:
@bparees so should we require a tag as well? eg.
```
$ openshift ex build-chain [image-repo]:[tag]
```

kargakis commented:
Also what should happen in the case of unspecified tags in buildConfigs? eg. when running sample-app i get:
```
[vagrant@openshiftdev sample-app]$ openshift cli get imageRepositoriesW0312 10:42:13.346463    1786 config.go:111] Unable to bind DNS as 0.0.0.0:53 (you may need to run as root), using 0.0.0.0:8053 which will not resolve from all locations
NAME                 DOCKER REPO                                     TAGS
origin-ruby-sample   172.30.17.223:5000/default/origin-ruby-sample   
ruby-20-centos7      172.30.17.223:5000/default/ruby-20-centos7 
```

No tags specified here. By requiring the user to specify tags in `build-chain`, we wouldn't get any results given one of these repositories. Maybe if the specified tag finds no match, then fall back to outputting all the repo dependencies regardless of tags?

smarterclayton commented:
Empty tag is "latest" and you should always interpret it that way (explicitly default it in code).


> On Mar 12, 2015, at 6:48 AM, Michail Kargakis <notifications@github.com> wrote:
> 
> Also what should happen in the case of unspecified tags in buildConfigs? eg. when running sample-app i get:
> 
> [vagrant@openshiftdev sample-app]$ openshift cli get imageRepositoriesW0312 10:42:13.346463    1786 config.go:111] Unable to bind DNS as 0.0.0.0:53 (you may need to run as root), using 0.0.0.0:8053 which will not resolve from all locations
> NAME                 DOCKER REPO                                     TAGS
> origin-ruby-sample   172.30.17.223:5000/default/origin-ruby-sample   
> ruby-20-centos7      172.30.17.223:5000/default/ruby-20-centos7 
> No tags specified here. By requiring the user to specify tags in build-chain, we wouldn't get any results given one of these repositories. Maybe if the specified tag finds no match, then fall back to outputting all the repo dependencies regardless of tags?
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
Users can omit tag, which means tag = "latest"


> On Mar 12, 2015, at 5:47 AM, Michail Kargakis <notifications@github.com> wrote:
> 
> @bparees so should we require a tag as well? eg.
> 
> $ openshift ex build-chain [image-repo]:[tag]
> —
> Reply to this email directly or view it on GitHub.
>

kargakis commented:
```
$ cd pkg/cmd/experimental/buildchain/
$ go test
# github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation
../../../../../../GoogleCloudPlatform/kubernetes/pkg/api/validation/schema.go:129: invalid operation: fieldDetails.DataTypeFields.Items[0] (type *swagger.Item does not support indexing)
FAIL    github.com/kargakis/origin/pkg/cmd/experimental/buildchain [build failed]
```

I jsut saw that the issue I was having with tests, got fixed last night upstream
https://github.com/GoogleCloudPlatform/kubernetes/commit/b2276b5d4b2dc926709d634e2c26bbe184a9413d
Tested the change and it fixes it. :)

smarterclayton commented:
I have a pull request that adds a generic graph processing behavior for builds, images, and image repositories that will allow us to better reflect large connections.  Later on (in a subsequent pull) we'll want to utilize that to back this function.


> On Mar 13, 2015, at 10:52 AM, Michail Kargakis <notifications@github.com> wrote:
> 
> $ cd pkg/cmd/experimental/buildchain/
> $ go test
> # github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation
> ../../../../../../GoogleCloudPlatform/kubernetes/pkg/api/validation/schema.go:129: invalid operation: fieldDetails.DataTypeFields.Items[0] (type *swagger.Item does not support indexing)
> FAIL    github.com/kargakis/origin/pkg/cmd/experimental/buildchain [build failed]
> I jsut saw that the issue I was having with tests, got fixed last night upstream
> GoogleCloudPlatform/kubernetes@b2276b5
> Tested the change and it fixes it. :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

kargakis commented:
@bparees @smarterclayton @csrwng mind reviewing?

All tests currently pass. In order to run them you have to go to your Kubernetes locally and add the change I mentioned above (just the one-liner in `schema.go`). The only thing that needs to be tested is setting up all those dummy buildConfigs inside OpenShift and running the tool so we can have complete testing of it with more than just two repositories (currently sample-app).

kargakis commented:
ps. sorry for the massive godeps commit, especially the parser adds a lot to it but we need it in order to valdiate the dot output.

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,324 @@
+	for _, cfg := range configs {
+		for _, tr := range cfg.Triggers {
+			if tr.ImageChange != nil {
+				avoidDuplicates[tr.ImageChange.From.Name] = struct{}{}
```
need to include From.Namespace if !nil in this uniqueness check

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,324 @@
+		for _, tr := range cfg.Triggers {
+			// If an ImageChangeTrigger has been specified (ie. it's not nil), use its From
+			// and Tag fields to compare against the specified image repository
+			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
```
needs to check ImageChange.From.Namespace if !nil

kargakis replied:
So if a user hasn't specified a namespace in the buildConfigs, they are invalid, right?

ps. From.Namespace is a string, so you must mean an empty string.

smarterclayton replied:
Namespace, if empty on an object reference, means "this namespace"


> On Mar 14, 2015, at 1:24 PM, Michail Kargakis <notifications@github.com> wrote:
> 
> In pkg/cmd/experimental/buildchain/buildchain.go:
> 
> > +func findRepoDeps(name, namespace, tag string, all bool, buildConfigList []buildapi.BuildConfig) *ImageRepo {
> > +	root := &ImageRepo{
> > +		Name:      name,
> > +		Namespace: namespace,
> > +		Tags:      []string{tag},
> > +	}
> > +
> > +	// Search all build configurations in order to find the image
> > +	// repositories depending on the specified image repository
> > +	var outputName, outputTag string
> > +	var err error
> > +	for _, cfg := range buildConfigList {
> > +		for _, tr := range cfg.Triggers {
> > +			// If an ImageChangeTrigger has been specified (ie. it's not nil), use its From
> > +			// and Tag fields to compare against the specified image repository
> > +			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
> So if a user hasn't specified a namespace in the buildConfigs, they are invalid, right?
> 
> ps. From.Namespace is a string, so you must mean an empty string.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bparees replied:
yes, empty string, sorry, java habits die hard.  if ImageChange.From.Namespace is populated, you need to compare it to the repo's namespace.  If imageChange.From.Namespace is not populated, then as @smarterclayton says, it means the namespace of the Build.

bparees replied:
why are you ignoring tag if all is true?  tag still needs to match.

bparees replied:
still not clear why you're ignoring the tag if --all is specified.  An ImageRepo update will never trigger an ICT if the tags do not match.

smarterclayton replied:
Yeah - tags must match.  We need to better expose methods that given an image repo and a tag, resolve it to verify a match.  That doesn't exist today, so default tag to "latest", and check equality.


> On Mar 17, 2015, at 1:04 PM, Ben Parees <notifications@github.com> wrote:
> 
> In pkg/cmd/experimental/buildchain/buildchain.go:
> 
> > +func findRepoDeps(name, namespace, tag string, all bool, buildConfigList []buildapi.BuildConfig) *ImageRepo {
> > +	root := &ImageRepo{
> > +		Name:      name,
> > +		Namespace: namespace,
> > +		Tags:      []string{tag},
> > +	}
> > +
> > +	// Search all build configurations in order to find the image
> > +	// repositories depending on the specified image repository
> > +	var outputName, outputTag string
> > +	var err error
> > +	for _, cfg := range buildConfigList {
> > +		for _, tr := range cfg.Triggers {
> > +			// If an ImageChangeTrigger has been specified (ie. it's not nil), use its From
> > +			// and Tag fields to compare against the specified image repository
> > +			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
> still not clear why you're ignoring the tag if --all is specified. An ImageRepo update will never trigger an ICT if the tags do not match.
> 
> —
> Reply to this email directly or view it on GitHub.
>

kargakis replied:
> An ImageRepo update will never trigger an ICT if the tags do not match.

By specifying a tag, we get back a specific tree, by using the --all flag we get back a tree constructed by using all the available tags in the root repository. I should probably check that tr.ImageChange.Tag != "" when the --all flag is specified.

Remember that if we let users specify tags with the --all flag then we get back to what you said about it on Trello.

kargakis replied:
> if ImageChange.From.Namespace is populated, you need to compare it to the repo's namespace.

This implies that the user has specified a namespace. So I guess if he/she doesn't, the default namespace will be used, right?

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,324 @@
+    $ openshift ex build-chain [image-repository]:[tag]
+
+    # Build dependency trees for all image repositories
+    $ openshift ex build-chain --all
```
did RCM specifically ask for --all?  it seems like it adds a fair bit of complexity and i'm not sure the output is really all that useful.  I might be inclined not to offer it.

smarterclayton replied:
I asked for it.

----- Original Message -----
> > +
> > +	buildapi "github.com/openshift/origin/pkg/build/api"
> > +	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
> > +	imageapi "github.com/openshift/origin/pkg/image/api"
> > +)
> > +
> > +const longDescription = `Output build dependencies of a specific image
> > repository.
> > +Supported output formats are json, dot, and ast. The default is set to
> > json.
> > +
> > +Examples:
> > +
> > +    # Build dependency tree for the specified image repository
> > +    $ openshift ex build-chain [image-repository]:[tag]
> > +
> > +    # Build dependency trees for all image repositories
> > +    $ openshift ex build-chain --all
> 
> did RCM specifically ask for --all?  it seems like it adds a fair bit of
> complexity and i'm not sure the output is really all that useful.  I might
> be inclined not to offer it.
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26411896

bparees replied:
what's the use case?  and are we going to output a tree per top level image repo, or a giant mesh?

bparees replied:
as it's currently implemented it only does --all for one namespace, should it really be doing it for all imagerepos in all namespaces the user has access to?

smarterclayton replied:
Yes - --all is very adminy.  Probably passing no argument should do all in the current namespace, and passing --all should work across all namespaces.

----- Original Message -----
> > +
> > +	buildapi "github.com/openshift/origin/pkg/build/api"
> > +	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
> > +	imageapi "github.com/openshift/origin/pkg/image/api"
> > +)
> > +
> > +const longDescription = `Output build dependencies of a specific image
> > repository.
> > +Supported output formats are json, dot, and ast. The default is set to
> > json.
> > +
> > +Examples:
> > +
> > +    # Build dependency tree for the specified image repository
> > +    $ openshift ex build-chain [image-repository]:[tag]
> > +
> > +    # Build dependency trees for all image repositories
> > +    $ openshift ex build-chain --all
> 
> as it's currently implemented it only does --all for one namespace, should it
> really be doing it for all imagerepos in all namespaces the user has access
> to?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26412826

kargakis replied:
> Yes - --all is very adminy. Probably passing no argument should do all in the current namespace, and passing --all should work across all namespaces.

+1

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,324 @@
+// type
+type Edge struct {
+	Name string `json:"buildConfig"`
+	To   string `json:"to"`
```
why not just make To a pointer to an ImageRepo?

kargakis replied:
Because in the end to produce json or dot output, we just need strings

kargakis commented:
The author of the gographviz library just exported it to [GitHub](https://github.com/awalterschulze/gographviz/commit/aa795c018b7bb2ef5ba889614b0cd72e6ee5dec7) (since Google Code [shuts down](google-opensource.blogspot.com/2015/03/farewell-to-google-code.html)) so I guess we better use that one as well.

smarterclayton commented:
https://github.com/openshift/origin/pull/1314 is the beginnings of a generalized graph traversal and library for api resources for all use cases, not just this.  Post 3.0 we'll want to come back and make this function merge into that, but it's not a requirement for this to be complete.

kargakis commented:
This is how the dot output is currently visualized
http://imgur.com/bhJrVjm
I cannot see the labels I have set up so I need to take a closer look into gographviz

mfojtik commented:
@kargakis do not copy the whole gographwiz repo into Godeps, just packages you need (so no README, LICENSE, etc...)

kargakis commented:
> @kargakis do not copy the whole gographwiz repo into Godeps, just packages you need (so no README, LICENSE, etc...)

Done

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+			var tag string
+			if !all && len(args) != 0 {
+				var repo string
+				repo, tag, err = breakDown(args[0])
```
just a nit... parse may be a better name than breakDown

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+				// since we are building dependency trees for every
+				// image repository available either in the current
+				// namespace or in every namespace
+				all = true
```
It may not be a good idea to assume all==true especially if you have a large repository. Unless they explicitly said that they wanted --all, I would warn that they need to specify an imagerepo

kargakis replied:
This will execute only in the cases of:
```
# output dep trees for all image repositories for all namespaces
$ openshift ex build-chain --all  
# output dep trees for all image repositories in the current namespace
$ openshift ex build-chain          
```

so essentially we silently make `--all` true only when no flag has been specified. The behavior of the utility from the point of setting `--all` to true is the same in both cases above.

csrwng replied:
got it

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+			if tr.ImageChange != nil {
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					avoidDuplicates[tr.ImageChange.From.Name] = defaultNamespace
```
Shouldn't the key to avoidDuplicates include the namespace? Otherwise you're going to only have one entry for the same name across namespaces

kargakis replied:
Good catch

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+					checkErr(fmt.Errorf("invalid image repository: %s", repo))
+				}
+
+				flushMap(repos)
```
wouldn't repos = {} accomplish the same thing?

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+			for _, namespace := range namespaces {
+				cfgList, err := osc.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+				checkErr(err)
+				for _, cfg := range cfgList.Items {
```
You can just do buildConfigList = append(buildConfigList, cfgList.Items...)

kargakis replied:
Sup

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+			// and Tag fields to compare against the specified image repository
+			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
+				// Either To & Tag or DockerImageReference will be used as output
+				if cfg.Parameters.Output.To.Name != "" {
```
cfg.Paramters.Output and cfg.Parameters.Output.To may be nil... should have:
if cfg.Parameters.Output != nil && cfg.Parameters.Output.To != nil && cfg.Parameters.Output.To.Name != ""

kargakis replied:
Output is not a pointer so I cannot compare it to nil. I guess comparing Output.To != nil is enough for both checks

csrwng replied:
Makes sense, for some reason at some point I thought we had made Output optional

smarterclayton replied:
Output is optional.

----- Original Message -----
> > +	root := &ImageRepo{
> > +		Name:      name,
> > +		Namespace: namespace,
> > +		Tags:      []string{tag},
> > +	}
> > +
> > +	// Search all build configurations in order to find the image
> > +	// repositories depending on the specified image repository
> > +	var outputName, outputTag string
> > +	for _, cfg := range buildConfigList {
> > +		for _, tr := range cfg.Triggers {
> > +			// If an ImageChangeTrigger has been specified (ie. it's not nil), use
> > its From
> > +			// and Tag fields to compare against the specified image repository
> > +			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all ||
> > tag == tr.ImageChange.Tag) {
> > +				// Either To & Tag or DockerImageReference will be used as output
> > +				if cfg.Parameters.Output.To.Name != "" {
> 
> Makes sense, for some reason at some point I thought we had made Output
> optional
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229/files#r26513419

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+				}
+
+				// Build all children and their dependency trees recursively
+				child := findRepoDeps(outputName, namespace, outputTag, all, buildConfigList)
```
It'd be good to keep a list of parents that you pass to findRepoDeps to avoid a cycle. Only add children that are not in the list of parents.

kargakis replied:
Yep, I already have thought about it and I agree with you. There is a Trello card about preventing the user from having buildConfigs with circular dependencies which will eliminate the usage of such a map but I guess I need to implement it right now

csrwng commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,346 @@
+		attrs["tag"+strconv.Itoa(i)] = strconv.Quote(tag)
+	}
+	attrs["namespace"] = strconv.Quote(root.Namespace)
+	rootName := strings.Replace(root.Name, "-", "_", -1)
```
is the hyphen("-") the only character that can be invalid in a dot node? If not, it may be good to abstract this into a function that turns a K8s name into a valid dot name.

kargakis replied:
I am not sure, will take a look at it

kargakis replied:
@csrwng 
```
An ID is one of the following:
* Any string of alphabetic ([a-zA-Z\200-\377]) characters, underscores ('_') or digits ([0-9]), not beginning with a digit;
* a numeral [-]?(.[0-9]+ | [0-9]+(.[0-9]*)? );
* any double-quoted string ("...") possibly containing escaped quotes (\")1;
* an HTML string (<...>). 
```
http://www.graphviz.org/doc/info/lang.html

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,339 @@
+			}
+
+			// Get all image repositories from build configurations
+			repos := getRepos(buildConfigList, defaultNamespace)
```
no reason to call this if "all" is not true.

bparees replied:
(and args is non-zero length)

kargakis replied:
I am using all the repos in order to validate that the specified image repository is valid (ie. part of the buildConfigs)

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,339 @@
+				}
+
+				// Build all children and their dependency trees recursively
+				child := findRepoDeps(namespace, outputName, outputTag, all, buildConfigList)
```
namespace here should be either cfg.Parameters.Output.To.Namespace(as you have) or cfg.Namespace.  As you have it now, the second case is going to use the previous imageRepo's namespace and not the buildConfig's namespace, which would be incorrect if the buildConfig is in a different namespace than the imagerepo that triggered the buildConfig.

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,339 @@
+				child := findRepoDeps(namespace, outputName, outputTag, all, buildConfigList)
+
+				// Add edge between root and child
+				root.Edges = append(root.Edges, NewEdge(cfg.Name, child.Name))
```
I think these names need to include the namespace as a prefix.

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,339 @@
+				if all {
+					cont := false
+					for _, repo := range root.Children {
+						if repo.Name == child.Name {
```
again needs to include namespace in de-duplication checking.

kargakis commented:
I fixed the labeling plus since we want to distinguish between namespaces I have added them in the name of a node (now called fullName). The new issue that pops up just by looking at the picture below is distinguishing between the name and the namespace:
http://i.imgur.com/oBLBnZD.png
The DOT parser eventually doesn't accept any special character except underscores...

kargakis commented:
An idea is to drop namespaces from DOT output altogether and use colors instead to distinguish between namespaces

smarterclayton commented:
Namespace will be important to most users 


> On Mar 17, 2015, at 8:26 AM, Michail Kargakis <notifications@github.com> wrote:
> 
> An idea is to drop namespaces from DOT output altogether and use colors instead to distinguish between namespaces
> 
> —
> Reply to this email directly or view it on GitHub.
>

kargakis commented:
@smarterclayton sure, take a look below where I output namespaces in smaller font
http://i.imgur.com/LVse6hX.png

smarterclayton commented:
Looks good

----- Original Message -----
> @smarterclayton sure, take a look below where I output namespaces in smaller
> font
> http://i.imgur.com/LVse6hX.png
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1229#issuecomment-82372804

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,333 @@
+			if tr.ImageChange != nil && tr.ImageChange.From.Name != "" {
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					repo = join(defaultNamespace, tr.ImageChange.From.Name)
```
Is defaultNamespace the right namespace to assign here or should I check first if cfg.Namespace is defined, if so, assign that, if not, assign defaultNamespace ? @bparees

kargakis replied:
Nvm, got my answer on IRC

kargakis commented:
Doc: https://github.com/openshift/openshift-docs/pull/225

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,331 @@
+					// TODO: Add a test that covers this path
+					ref, err := imageapi.ParseDockerImageReference(cfg.Parameters.Output.DockerImageReference)
+					checkErr(err)
+					outputName = ref.Name
```
Is ref.Name the name of the image or the image repository? @bparees

kargakis replied:
Nvm v.2, it's the image repo name

kargakis commented:
I think this is ready for one more (hopefully the last) review:)

cc: @bparees @csrwng @smarterclayton

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,368 @@
+				// Validate the specified image repository
+				repo := join(namespace, name)
+				if _, ok := repos[repo]; !ok {
+					checkErr(fmt.Errorf("invalid image repository: %s", repo))
```
it isn't really invalid, it's just an image repo that doesn't trigger any builds, right?

kargakis replied:
Well it may be that or it may be invalid

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,367 @@
+
+// getRepos iterates over a given set of build configurations
+// and extracts all the image repositories on which other
+// image repositories are based
```
really this is "extracts all the imagerepos which trigger a buildconfig when the imagerepo is updated"

bparees commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,367 @@
+					}
+					childName = ref.Name
+					childTag = ref.Tag
+					childNamespace = cfg.Namespace
```
i don't think we need to change this, but it's worth noting that the imagereference's namespace (eg "openshift" in a pullspec like openshift/ruby-20-centos7:foo) could actually be the namespace of the imagerepo.  unfortunately it's impossible to know, so using the buildconfig's namespace is just as reasonable.

fundamentally if your buildconfig outputs to an imageref instead of an imagerepo, things get pretty fuzzy.

bparees commented:
lgtm and it's experimental so we can iterate a little if we need to.  great work @kargakis 

let me know if you want it merged, wasn't sure if you had any final tweaks or that's just doc updates.

kargakis commented:
> let me know if you want it merged, wasn't sure if you had any final tweaks or that's just doc updates.

@bparees I want to add tags on node names for the DOT output and was thinking about adding an extra flag where we would output all the trees generated by a specific repository which we currently don't support and it's probably something that the RCM guys will probably want.

eg.
```
$ openshift ex ruby-20-centos7 --complete
```
generates all trees from all tags in this repo.

Thinking about it, we support outputing all trees but for all image repos and not for one specific.
What do you think? Also any suggestions about the name of the new flag?
cc: @smarterclayton @csrwng

bparees commented:
@karkgakis sounds reasonable. Maybe "--all-tags"

kargakis commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -0,0 +1,367 @@
+				// If the user specifies the --all flag and the child depends
+				// on root via more than one tag, multiple instances of the child
+				// will make it into root.Children
+				if all {
```
@bparees it's possible that two buildConfigs will have the same From.ICT.Tag (and same repo ofc) but two different Output.Tags of another single repository right? If so, I think this all check is wrong and this should be checked every time.

kargakis commented:
@bparees have a look in the latest commits and I'll squash if everything is fine. I won't add tags next to names for the time being because it gets complicated with children with multiple tags.

kargakis commented:
I have injected tags into root names (only) because I feel it's better to show root tags than not
http://i.imgur.com/2GxyL8d.png

bparees commented:
@kargakis still lgtm.

kargakis commented:
Cool, you can merge it now

bparees commented:
[merge]
